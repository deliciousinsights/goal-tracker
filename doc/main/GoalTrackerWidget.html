<!DOCTYPE html><html lang="en"><head><title>main/GoalTrackerWidget</title></head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0"><meta name="groc-relative-root" content="../"><meta name="groc-document-path" content="main/GoalTrackerWidget"><meta name="groc-project-path" content="src/main/GoalTrackerWidget.js"><link rel="stylesheet" type="text/css" media="all" href="../assets/style.css"><script type="text/javascript" src="../assets/behavior.js"></script><body><div id="meta"><div class="file-path">src/main/GoalTrackerWidget.js</div></div><div id="document"><div class="segment"><div class="comments "><div class="wrapper"><h1 id="suivi-du-jour-pour-un-objectif">Suivi du jour pour un objectif</h1></div></div></div><div class="segment"><div class="code"><div class="wrapper"><span class="hljs-keyword">import</span> Add <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/icons-material/Add'</span>
<span class="hljs-keyword">import</span> Fab <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/material/Fab'</span>
<span class="hljs-keyword">import</span> ThumbUp <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/icons-material/ThumbUp'</span>
<span class="hljs-keyword">import</span> Typography <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/material/Typography'</span>

<span class="hljs-keyword">import</span> classes <span class="hljs-keyword">from</span> <span class="hljs-string">'./TrackerScreen.module.css'</span>
<span class="hljs-keyword">import</span> { func, GoalPropType, nonNegativeInteger } <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/prop-types'</span>
<span class="hljs-keyword">import</span> Gauge <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/Gauge'</span></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Section de l&#39;écran principal, dédiée à un objectif.  Fournit notamment le
descriptif de l&#39;objectif et l’éventuel bouton de progression.</p></div></div><div class="code"><div class="wrapper"><span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> <span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">GoalTrackerWidget</span>(<span class="hljs-params">{</span></span></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>La déstructuration en force !</p></div></div><div class="code"><div class="wrapper">  goal,
  goal: { name, units, target },
  onProgress,
  progress,
}) {</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>La beauté d&#39;un ternaire multi-lignes…</p></div></div><div class="code"><div class="wrapper">  <span class="hljs-keyword">const</span> adderComponent =
    target &gt; progress ? (
      <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">Fab</span>
        <span class="hljs-attribute">color</span>=<span class="hljs-value">'secondary'</span>
        <span class="hljs-attribute">size</span>=<span class="hljs-value">'small'</span>
        <span class="hljs-attribute">aria-label</span>=<span class="hljs-value">{`Progresser</span> <span class="hljs-attribute">sur</span> ${<span class="hljs-attribute">name</span>}`}
        <span class="hljs-attribute">onClick</span>=<span class="hljs-value">{()</span> =&gt;</span> onProgress?.(goal)}
      &gt;
        <span class="hljs-tag">&lt;<span class="hljs-title">Add</span> <span class="hljs-attribute">data-testid</span>=<span class="hljs-value">'in-progress'</span> /&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-title">Fab</span>&gt;</span>
    )</span> : (
      <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">Fab</span> <span class="hljs-attribute">disabled</span> <span class="hljs-attribute">size</span>=<span class="hljs-value">'small'</span> <span class="hljs-attribute">aria-label</span>=<span class="hljs-value">'Objectif atteint, bravo !'</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">ThumbUp</span> <span class="hljs-attribute">data-testid</span>=<span class="hljs-value">'completed'</span> /&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-title">Fab</span>&gt;</span>
    )</span>

  <span class="hljs-keyword">return</span> (
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">div</span> <span class="hljs-attribute">className</span>=<span class="hljs-value">{classes.goal}</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-title">div</span> <span class="hljs-attribute">className</span>=<span class="hljs-value">{classes.summary}</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">Typography</span> <span class="hljs-attribute">variant</span>=<span class="hljs-value">'h6'</span> <span class="hljs-attribute">component</span>=<span class="hljs-value">'h2'</span>&gt;</span>
          {name}
        <span class="hljs-tag">&lt;/<span class="hljs-title">Typography</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">Gauge</span> <span class="hljs-attribute">value</span>=<span class="hljs-value">{progress}</span> <span class="hljs-attribute">max</span>=<span class="hljs-value">{target}</span> /&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">Typography</span> <span class="hljs-attribute">component</span>=<span class="hljs-value">'small'</span>&gt;</span>
          {`${progress} ${units} sur ${target}`}
        <span class="hljs-tag">&lt;/<span class="hljs-title">Typography</span>&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-title">div</span> <span class="hljs-attribute">className</span>=<span class="hljs-value">{classes.cta}</span>&gt;</span>{adderComponent}<span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
    <span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
  )</span>
}

GoalTrackerWidget.propTypes = {
  goal: GoalPropType.isRequired,
  progress: nonNegativeInteger.isRequired,
  onProgress: func,
}</div></div></div></div></body></html>