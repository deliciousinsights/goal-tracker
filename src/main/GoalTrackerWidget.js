import Add from '@mui/icons-material/Add'
import Fab from '@mui/material/Fab'
import ThumbUp from '@mui/icons-material/ThumbUp'
import Typography from '@mui/material/Typography'

import classes from './TrackerScreen.module.css'
import { func, GoalPropType, nonNegativeInteger } from '../shared/prop-types'
import Gauge from '../shared/Gauge'

export default function GoalTrackerWidget({
  goal,
  goal: { name, units, target },
  onProgress,
  progress,
}) {
  const adderComponent =
    target > progress ? (
      <Fab
        color='secondary'
        size='small'
        aria-label={`Progresser sur ${name}`}
        onClick={() => onProgress?.(goal)}
      >
        <Add data-testid='in-progress' />
      </Fab>
    ) : (
      <Fab disabled size='small' aria-label='Objectif atteint, bravoÂ !'>
        <ThumbUp data-testid='completed' />
      </Fab>
    )

  return (
    <div className={classes.goal}>
      <div className={classes.summary}>
        <Typography variant='h6' component='h2'>
          {name}
        </Typography>
        <Gauge value={progress} max={target} />
        <Typography component='small'>
          {`${progress} ${units} sur ${target}`}
        </Typography>
      </div>
      <div className={classes.cta}>{adderComponent}</div>
    </div>
  )
}

GoalTrackerWidget.propTypes = {
  goal: GoalPropType.isRequired,
  progress: nonNegativeInteger.isRequired,
  onProgress: func,
}
