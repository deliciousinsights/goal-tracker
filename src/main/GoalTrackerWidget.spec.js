import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import GoalTrackerWidget from './GoalTrackerWidget'

describe('<GoalTrackerWidget />', () => {
  const goal = {
    id: '0123456789abcdef01234567',
    name: 'My goal',
    target: 42,
    units: 'wombats',
  }
  describe('when not completed', () => {
    it.each`
      progress | percentage
      ${0}     | ${0}
      ${1}     | ${2}
      ${21}    | ${50}
      ${41}    | ${98}
    `(
      'should render appropriately at progress $progress',
      ({ progress, percentage }) => {
        render(<GoalTrackerWidget goal={goal} progress={progress} />)
        expect(screen.getByRole('heading')).toHaveTextContent(goal.name)
        expect(screen.getByRole('progressbar')).toHaveAttribute(
          'aria-valuenow',
          String(percentage)
        )
        expect(
          screen.getByText(`${progress} ${goal.units} sur ${goal.target}`)
        ).toBeInTheDocument()

        expect(screen.getByTestId('in-progress')).toBeInTheDocument()
      }
    )

    it('should trigger its onProgress on click', async () => {
      const progress = 21
      const onProgress = jest.fn()
      render(
        <GoalTrackerWidget
          goal={goal}
          onProgress={onProgress}
          progress={progress}
        />
      )

      await userEvent.click(screen.getByRole('button'))
      expect(onProgress).toHaveBeenCalledTimes(1)
      expect(onProgress).toHaveBeenCalledWith(goal)
    })

    it('should otherwise match the expected snapshot', () => {
      const { container } = render(
        <GoalTrackerWidget goal={goal} progress={21} />
      )

      expect(container).toMatchSnapshot()
    })
  })

  describe('when completed (or exceeded)', () => {
    it.each([goal.target, goal.target + 1, goal.target + 10])(
      'should render appropriately at progress %i',
      (progress) => {
        render(<GoalTrackerWidget goal={goal} progress={progress} />)

        expect(screen.queryByTestId('in-progress')).not.toBeInTheDocument()
        expect(screen.getByTestId('completed')).toBeInTheDocument()
      }
    )

    it('should otherwise match the expected snapshot', () => {
      const { container } = render(
        <GoalTrackerWidget goal={goal} progress={42} />
      )

      expect(container).toMatchSnapshot()
    })
  })
})
