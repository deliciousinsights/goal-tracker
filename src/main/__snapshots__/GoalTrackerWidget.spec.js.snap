// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoalTrackerWidget /> when completed (or exceeded) should otherwise match the expected snapshot 1`] = `
<div>
  <div
    class="goal"
  >
    <div
      class="summary"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        My goal
      </h2>
      <span
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="100"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
        style="height: 8px;"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(0%);"
        />
      </span>
      <small
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        42 wombats sur 42
      </small>
    </div>
    <div
      class="cta"
    >
      <button
        aria-label="Objectif atteint, bravoÂ !"
        class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default Mui-disabled MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default css-12m9wit-MuiButtonBase-root-MuiFab-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="completed"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`<GoalTrackerWidget /> when not completed should otherwise match the expected snapshot 1`] = `
<div>
  <div
    class="goal"
  >
    <div
      class="summary"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        My goal
      </h2>
      <span
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="50"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
        style="height: 8px;"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(-50%);"
        />
      </span>
      <small
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        21 wombats sur 42
      </small>
    </div>
    <div
      class="cta"
    >
      <button
        aria-label="Progresser sur My goal"
        class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary css-b7losd-MuiButtonBase-root-MuiFab-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="in-progress"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
